tests/test_algorithm/test_idcd.py

```diff
@@ -0,0 +1,455 @@
+            directed=[
+                (X, Y),
+                (Y, X),
+                (X, Z),
+            ]
+        )
+
+        targets = {Z}
+        ancestral_closure = {X, Y, Z}
+
+        with self.assertRaises(Unidentifiable) as context:
+            identify_through_scc_decomposition(
+                graph=graph,
+                targets=targets,
+                ancestral_closure=ancestral_closure,
+                recursion_level=0,
+            )
+
+        self.assertIn("No SCCs", str(context.exception))
+
+
+# ----------------------------------------------------------------------------
+
+
+class TestComputeSCCDistributions(unittest.TestCase):
+    """Tests for compute_scc_distributions function."""
+
+    def test_single_scc_returns_correct_structure(self) -> None:
+        """Test that a single SCC returns the correct distribution structure."""
+        graph = NxMixedGraph.from_edges(
+            directed=[
+                (X, Y),
+                (Y, Z),
+                (Z, X),
+            ]
+        )
+
+        subgraph_a = graph.subgraph({X, Y, Z})
+
+        relevant_sccs = [frozenset({X, Y, Z})]
+        ancestral_closure = {X, Y, Z}
+
+        result = compute_scc_distributions(
+            graph=graph,
+            subgraph_a=subgraph_a,
+            relevant_sccs=relevant_sccs,
+            ancestral_closure=ancestral_closure,
+        )
+
+        # Should return a dictionary
+        self.assertIsInstance(result, dict)
+
+        # Should have one entry (one SCC)
+        self.assertEqual(len(result), 1)
+
+        # The key should be the SCC
+        self.assertIn(frozenset({X, Y, Z}), result)
+
+        # The value should be an Expression
+        distribution = result[frozenset({X, Y, Z})]
+        self.assertIsInstance(distribution, Expression)
+
+    def test_multiple_sccs_with_cycles(self) -> None:
+        """Test multiple SCCs each with cycles in the graph."""
+        graph = NxMixedGraph.from_edges(
+            directed=[
+                (X, X),
+                (Y, X),
+                (W, Z),
+                (Z, W),
+            ]
+        )
+
+        # subgraph contains both cycles in the graph
+        subgraph_a = graph.subgraph({X, Y, W, Z})
+
+        # two SCCs
+        relevant_sccs = [frozenset({X, Y}), frozenset({W, Z})]
+        ancestral_closure = {X, Y, W, Z}
+
+        result = compute_scc_distributions(
+            graph=graph,
+            subgraph_a=subgraph_a,
+            relevant_sccs=relevant_sccs,
+            ancestral_closure=ancestral_closure,
+        )
+
+        # Should return a dictionary
+        self.assertIsInstance(result, dict)
+
+        # should have two entries (two SCCs)
+        self.assertEqual(len(result), 2)
+
+        # Each key should be an SCC
+        self.assertIn(frozenset({X, Y}), result)
+        self.assertIn(frozenset({W, Z}), result)
+
+        for scc in relevant_sccs:
+            self.assertIsInstance(result[scc], Expression)
```

Comment:
you need to do better than this for unit tests ;p

what are the expressions? write them out in full to make sure we have a full understanding of what the input/output should be